README.md
<img src="https://raw.githubusercontent.com/mirnovov/colourpicker/main/colourpicker.gif" style="border-radius: 12px;" />

**NOTE**: As Nova 5 and above includes a native colour picker, this extension will not receive further updates.

The **Colour Picker** Nova extension allows you to insert colours into your HTML, CSS, and Javascript files using the default macOS colour selection dialog. 

### Usage

Use the *Choose Colour* (```⌘*```) right-click command to choose a colour and insert it into the text.

If a colour is selected, the picker shows the existing colour, and overwrites it using the same format. Otherwise, the format is determined by the *Default Colour Format* preferences item. 

By default, hex colours are generated in lowercase. Select the *Capitalise generated hex codes* preference to modify this.

### Limitations

This extension is implemented using JXA and the ```osxascript``` process. Accordingly, modification of alpha channels is not supported.

CHANGELOG.md
* **Version 0.1**: Initial release
* **Version 0.2**: Fix missing image asset
* **Version 0.3**: Enabled usage regardless of editor syntax selected
* **Version 1.0**:
	* Add CSS named colour support
	* Add support for P3 colours using the ``color()`` CSS Level 4 function
	* Improved performance by switching to JXA backend
	* Improved precision during edge cases
* **Version 1.0.1**: Removed unnecessary entitlement
* **Version 1.0.2**: Add repo details
* **Version 1.0.3**: Fix links
* **Version 1.0.4**: Added depreciation notice

Scripts/namedcolours.js
export var namedList = {
    "aliceblue": "0.9411764705882353,0.9725490196078431,1",
    "antiquewhite": "0.9803921568627451,0.9215686274509803,0.8431372549019608",
    "aqua": "0,1,1",
    "aquamarine": "0.4980392156862745,1,0.8313725490196079",
    "azure": "0.9411764705882353,1,1",
    "beige": "0.9607843137254902,0.9607843137254902,0.8627450980392157",
    "bisque": "1,0.8941176470588236,0.7686274509803922",
    "black": "0,0,0",
    "blanchedalmond": "1,0.9215686274509803,0.803921568627451",
    "blue": "0,0,1",
    "blueviolet": "0.5411764705882353,0.16862745098039217,0.8862745098039215",
    "brown": "0.6470588235294118,0.16470588235294117,0.16470588235294117",
    "burlywood": "0.8705882352941177,0.7215686274509804,0.5294117647058824",
    "cadetblue": "0.37254901960784315,0.6196078431372549,0.6274509803921569",
    "chartreuse": "0.4980392156862745,1,0",
    "chocolate": "0.8235294117647058,0.4117647058823529,0.11764705882352941",
    "coral": "1,0.4980392156862745,0.3137254901960784",
    "cornflowerblue": "0.39215686274509803,0.5843137254901961,0.9294117647058824",
    "cornsilk": "1,0.9725490196078431,0.8627450980392157",
    "crimson": "0.8627450980392157,0.0784313725490196,0.23529411764705882",
    "cyan": "0,1,1",
    "darkblue": "0,0,0.5450980392156862",
    "darkcyan": "0,0.5450980392156862,0.5450980392156862",
    "darkgoldenrod": "0.7215686274509804,0.5254901960784314,0.043137254901960784",
    "darkgray": "0.6627450980392157,0.6627450980392157,0.6627450980392157",
    "darkgrey": "0.6627450980392157,0.6627450980392157,0.6627450980392157",
    "darkgreen": "0,0.39215686274509803,0",
    "darkkhaki": "0.7411764705882353,0.7176470588235294,0.4196078431372549",
    "darkmagenta": "0.5450980392156862,0,0.5450980392156862",
    "darkolivegreen": "0.3333333333333333,0.4196078431372549,0.1843137254901961",
    "darkorange": "1,0.5490196078431373,0",
    "darkorchid": "0.6,0.19607843137254902,0.8",
    "darkred": "0.5450980392156862,0,0",
    "darksalmon": "0.9137254901960784,0.5882352941176471,0.47843137254901963",
    "darkseagreen": "0.5607843137254902,0.7372549019607844,0.5607843137254902",
    "darkslateblue": "0.2823529411764706,0.23921568627450981,0.5450980392156862",
    "darkslategray": "0.1843137254901961,0.30980392156862746,0.30980392156862746",
    "darkslategrey": "0.1843137254901961,0.30980392156862746,0.30980392156862746",
    "darkturquoise": "0,0.807843137254902,0.8196078431372549",
    "darkviolet": "0.5803921568627451,0,0.8274509803921568",
    "deeppink": "1,0.0784313725490196,0.5764705882352941",
    "deepskyblue": "0,0.7490196078431373,1",
    "dimgray": "0.4117647058823529,0.4117647058823529,0.4117647058823529",
    "dimgrey": "0.4117647058823529,0.4117647058823529,0.4117647058823529",
    "dodgerblue": "0.11764705882352941,0.5647058823529412,1",
    "firebrick": "0.6980392156862745,0.13333333333333333,0.13333333333333333",
    "floralwhite": "1,0.9803921568627451,0.9411764705882353",
    "forestgreen": "0.13333333333333333,0.5450980392156862,0.13333333333333333",
    "fuchsia": "1,0,1",
    "gainsboro": "0.8627450980392157,0.8627450980392157,0.8627450980392157",
    "ghostwhite": "0.9725490196078431,0.9725490196078431,1",
    "gold": "1,0.8431372549019608,0",
    "goldenrod": "0.8549019607843137,0.6470588235294118,0.12549019607843137",
    "gray": "0.5019607843137255,0.5019607843137255,0.5019607843137255",
    "grey": "0.5019607843137255,0.5019607843137255,0.5019607843137255",
    "green": "0,0.5019607843137255,0",
    "greenyellow": "0.6784313725490196,1,0.1843137254901961",
    "honeydew": "0.9411764705882353,1,0.9411764705882353",
    "hotpink": "1,0.4117647058823529,0.7058823529411765",
    "indianred": "0.803921568627451,0.3607843137254902,0.3607843137254902",
    "indigo": "0.29411764705882354,0,0.5098039215686274",
    "ivory": "1,1,0.9411764705882353",
    "khaki": "0.9411764705882353,0.9019607843137255,0.5490196078431373",
    "lavender": "0.9019607843137255,0.9019607843137255,0.9803921568627451",
    "lavenderblush": "1,0.9411764705882353,0.9607843137254902",
    "lawngreen": "0.48627450980392156,0.9882352941176471,0",
    "lemonchiffon": "1,0.9803921568627451,0.803921568627451",
    "lightblue": "0.6784313725490196,0.8470588235294118,0.9019607843137255",
    "lightcoral": "0.9411764705882353,0.5019607843137255,0.5019607843137255",
    "lightcyan": "0.8784313725490196,1,1",
    "lightgoldenrodyellow": "0.9803921568627451,0.9803921568627451,0.8235294117647058",
    "lightgray": "0.8274509803921568,0.8274509803921568,0.8274509803921568",
    "lightgrey": "0.8274509803921568,0.8274509803921568,0.8274509803921568",
    "lightgreen": "0.5647058823529412,0.9333333333333333,0.5647058823529412",
    "lightpink": "1,0.7137254901960784,0.7568627450980392",
    "lightsalmon": "1,0.6274509803921569,0.47843137254901963",
    "lightseagreen": "0.12549019607843137,0.6980392156862745,0.6666666666666666",
    "lightskyblue": "0.5294117647058824,0.807843137254902,0.9803921568627451",
    "lightslategray": "0.4666666666666667,0.5333333333333333,0.6",
    "lightslategrey": "0.4666666666666667,0.5333333333333333,0.6",
    "lightsteelblue": "0.6901960784313725,0.7686274509803922,0.8705882352941177",
    "lightyellow": "1,1,0.8784313725490196",
    "lime": "0,1,0",
    "limegreen": "0.19607843137254902,0.803921568627451,0.19607843137254902",
    "linen": "0.9803921568627451,0.9411764705882353,0.9019607843137255",
    "magenta": "1,0,1",
    "maroon": "0.5019607843137255,0,0",
    "mediumaquamarine": "0.4,0.803921568627451,0.6666666666666666",
    "mediumblue": "0,0,0.803921568627451",
    "mediumorchid": "0.7294117647058823,0.3333333333333333,0.8274509803921568",
    "mediumpurple": "0.5764705882352941,0.4392156862745098,0.8470588235294118",
    "mediumseagreen": "0.23529411764705882,0.7019607843137254,0.44313725490196076",
    "mediumslateblue": "0.4823529411764706,0.40784313725490196,0.9333333333333333",
    "mediumspringgreen": "0,0.9803921568627451,0.6039215686274509",
    "mediumturquoise": "0.2823529411764706,0.8196078431372549,0.8",
    "mediumvioletred": "0.7803921568627451,0.08235294117647059,0.5215686274509804",
    "midnightblue": "0.09803921568627451,0.09803921568627451,0.4392156862745098",
    "mintcream": "0.9607843137254902,1,0.9803921568627451",
    "mistyrose": "1,0.8941176470588236,0.8823529411764706",
    "moccasin": "1,0.8941176470588236,0.7098039215686275",
    "navajowhite": "1,0.8705882352941177,0.6784313725490196",
    "navy": "0,0,0.5019607843137255",
    "oldlace": "0.9921568627450981,0.9607843137254902,0.9019607843137255",
    "olive": "0.5019607843137255,0.5019607843137255,0",
    "olivedrab": "0.4196078431372549,0.5568627450980392,0.13725490196078433",
    "orange": "1,0.6470588235294118,0",
    "orangered": "1,0.27058823529411763,0",
    "orchid": "0.8549019607843137,0.4392156862745098,0.8392156862745098",
    "palegoldenrod": "0.9333333333333333,0.9098039215686274,0.6666666666666666",
    "palegreen": "0.596078431372549,0.984313725490196,0.596078431372549",
    "paleturquoise": "0.6862745098039216,0.9333333333333333,0.9333333333333333",
    "palevioletred": "0.8470588235294118,0.4392156862745098,0.5764705882352941",
    "papayawhip": "1,0.9372549019607843,0.8352941176470589",
    "peachpuff": "1,0.8549019607843137,0.7254901960784313",
    "peru": "0.803921568627451,0.5215686274509804,0.24705882352941178",
    "pink": "1,0.7529411764705882,0.796078431372549",
    "plum": "0.8666666666666667,0.6274509803921569,0.8666666666666667",
    "powderblue": "0.6901960784313725,0.8784313725490196,0.9019607843137255",
    "purple": "0.5019607843137255,0,0.5019607843137255",
    "red": "1,0,0",
    "rebeccapurple": "0.4,0.2,0.6",
    "rosybrown": "0.7372549019607844,0.5607843137254902,0.5607843137254902",
    "royalblue": "0.2549019607843137,0.4117647058823529,0.8823529411764706",
    "saddlebrown": "0.5450980392156862,0.27058823529411763,0.07450980392156863",
    "salmon": "0.9803921568627451,0.5019607843137255,0.4470588235294118",
    "sandybrown": "0.9568627450980393,0.6431372549019608,0.3764705882352941",
    "seagreen": "0.1803921568627451,0.5450980392156862,0.3411764705882353",
    "seashell": "1,0.9607843137254902,0.9333333333333333",
    "sienna": "0.6274509803921569,0.3215686274509804,0.17647058823529413",
    "silver": "0.7529411764705882,0.7529411764705882,0.7529411764705882",
    "skyblue": "0.5294117647058824,0.807843137254902,0.9215686274509803",
    "slateblue": "0.41568627450980394,0.35294117647058826,0.803921568627451",
    "slategray": "0.4392156862745098,0.5019607843137255,0.5647058823529412",
    "slategrey": "0.4392156862745098,0.5019607843137255,0.5647058823529412",
    "snow": "1,0.9803921568627451,0.9803921568627451",
    "springgreen": "0,1,0.4980392156862745",
    "steelblue": "0.27450980392156865,0.5098039215686274,0.7058823529411765",
    "tan": "0.8235294117647058,0.7058823529411765,0.5490196078431373",
    "teal": "0,0.5019607843137255,0.5019607843137255",
    "thistle": "0.8470588235294118,0.7490196078431373,0.8470588235294118",
    "tomato": "1,0.38823529411764707,0.2784313725490196",
    "turquoise": "0.25098039215686274,0.8784313725490196,0.8156862745098039",
    "violet": "0.9333333333333333,0.5098039215686274,0.9333333333333333",
    "wheat": "0.9607843137254902,0.8705882352941177,0.7019607843137254",
    "white": "1,1,1",
    "whitesmoke": "0.9607843137254902,0.9607843137254902,0.9607843137254902",
    "yellow": "1,1,0",
    "yellowgreen": "0.6039215686274509,0.803921568627451,0.19607843137254902"
};



Scripts/main.js
exports.activate = function() {
    nova.commands.register("novov.colourpicker.changeColour", (editor) => {
        var oldColour = interpretColour(editor.getTextInRange(editor.selectedRanges[0]));
        var process = new Process("/usr/bin/env", {
            args: ["osascript", "AppleScript/colourpickerjs.scpt",oldColour["r"],oldColour["g"],oldColour["b"]],
            cwd: nova.extension.path,
            stdio: ["ignore", "pipe", "pipe"],
        }); 
        
        process.onStdout(function (newColour) { 
            if(newColour != null) { 
                editor.insert(formatColour(newColour,oldColour["mode"],oldColour["alpha"]) + oldColour["end"]); 
            }
        });
        process.start(); 
    });
}

function interpretColour(text) {
    const hexmatch = "#([A-Fa-f0-9]{3,8})",
          rgbmatch = "(rgba|rgb)\\(([\\d|\\.]+),([\\d|\\.]+),([\\d|\\.]+),?([\\d|\\.]*)\\)",
          hslmatch = "(hsl|hsla)\\(([\\d|\\.]+),([\\d|\\.]+),([\\d|\\.]+),?([\\d|\\.]*)\\)",
          ptrmatch = "color\\(display-p3 ([\\d|\\.]+) ([\\d|\\.]+) ([\\d|\\.]+) ?([\\d|\\.]*)\\)";
    
    var mode = nova.config.get("novov.colourpicker.default").toLowerCase(),
        re = text.match("#([A-Fa-f0-9]{3,8})"),
        r = 1, g = 1, b = 1,
        alpha = null,
        end = "";
        
    if(re = text.match(hexmatch)) {
        mode = "hex"; 
        var hex = re[1];
        
        if(text.length > 5) {
            r = dehex(hex.slice(0,2),255);
            g = dehex(hex.slice(2,4),255);
            b = dehex(hex.slice(4,6),255);
        }
        else {
            r = dehex(hex.slice(0,1),15);
            g = dehex(hex.slice(1,2),15);
            b = dehex(hex.slice(2,3),15);
        }
        
        if (hex.length > 7) { alpha = hex.slice(6,8); }
        else if (hex.length == 4) { alpha = hex.slice(3,4).repeat(2); }
    }
    else if(re = text.match(rgbmatch)) {
        mode = "rgb"; 
        r = parseInt(re[2]) / 255;
        g = parseInt(re[3]) / 255;
        b = parseInt(re[4]) / 255;
        alpha = re[5] == "" ? null : parseInt(re[5]);
    }
    else if(re = text.match(hslmatch)) {
        mode = "hsl"; 
        [r, g, b] = HSLToRGB(parseInt(re[2]), parseInt(re[3]), parseInt(re[4]));
        alpha = re[5] == "" ? null : parseInt(re[5]);
    }
    else if(re = text.match(ptrmatch)) {
        mode = "ptr"; 
        r = parseFloat(re[1]), g = parseFloat(re[2]), b = parseFloat(re[3]);
        alpha = re[4] == "" ? null : parseInt(re[4]);
    }
    else if(text == "transparent") {
        r = 0, g = 0, b = 0;
        switch(mode) {
            case "hex": alpha = "00"; break;
            default: alpha = "0"; break;
        }
    }
    else if(re = namedColours()[text.toLowerCase()]) {
        var value = re.split(",").map(parseFloat);
        r = value[0], g = value[1], b = value[2];
    }
    
    if(text.endsWith("\n")) { end = "\n"; }
    if(mode == "Display P3") { mode = "ptr"; }
        
    return {
        "mode": mode,
        "r": r.toString(),
        "g": g.toString(),
        "b": b.toString(),
        "alpha": alpha,
        "end": end  
    };
}

function dehex(hex, divisor) {
    return parseInt("0x" + hex, 16) / divisor;
}

function formatColour(colour,format,alpha) {
    var hues = colour.split(", ");
    
    switch (format) {
      case "ptr":
        hues = hues.map(function(i) { return parseFloat(i).toFixed(1); }); 
        break;
      case "hsl":
        hues = hues.map(parseFloat); 
        break;
      default:
        hues = hues.map(function(i) { return Math.round(parseFloat(i) * 255); }); 
        break;
    }
    
    switch (format) {
      case "hex":
        for(var i = 0; i < hues.length; i++) {
            hues[i] = hues[i].toString(16);
            if(hues[i].length == 1) { hues[i] = "0" + hues[i]; }
        }
        return formatHex(`#${hues[0]}${hues[1]}${hues[2]}${alpha != null ? alpha : ""}`);
      case "rgb":
        if(alpha != null) { return `rgba(${hues[0]},${hues[1]},${hues[2]},${alpha})`; }
        return `rgb(${hues[0]},${hues[1]},${hues[2]})`;
      case "ptr":
        if(alpha != null) { return `color(display-p3 ${hues[0]} ${hues[1]} ${hues[2]} ${alpha})`; }
        return `color(display-p3 ${hues[0]} ${hues[1]} ${hues[2]})`;
      case "hsl":
        return RGBToHSL(hues[0],hues[1],hues[2],alpha);
    }
}

function formatHex(hex) {
    if (nova.config.get("novov.colourpicker.capitalise")) { return hex.toUpperCase(); }
    return hex;
}

//Two functions below thanks to https://css-tricks.com/converting-color-spaces-in-javascript/
function RGBToHSL(r,g,b,alpha) {
    var cmin = Math.min(r,g,b),
        cmax = Math.max(r,g,b),
        delta = cmax - cmin,
        h = 0,
        s = 0,
        l = 0;
    
    if (delta == 0) { h = 0; }
    else if (cmax == r) { h = ((g - b) / delta) % 6; }
    else if (cmax == g) { h = (b - r) / delta + 2; }
    else { h = (r - g) / delta + 4; }
    
    h = Math.round(h * 60);
    if (h < 0) { h += 360; }
        
    l = (cmax + cmin) / 2;
    s = delta == 0 ? 0 : delta / (1 - Math.abs(2 * l - 1));
    s = +(s * 100).toFixed(1);
    l = +(l * 100).toFixed(1);
    
    if(alpha != null) { return `hsla(${h},${s},${l},${alpha})` }
    return `hsl(${h},${s},${l})`
}

function HSLToRGB(h,s,l) {
    s /= 100;
    l /= 100;
  
    var c = (1 - Math.abs(2 * l - 1)) * s,
        x = c * (1 - Math.abs((h / 60) % 2 - 1)),
        m = l - c/2,
        r = 0,
        g = 0,
        b = 0;

    if (0 <= h && h < 60) { r = c; g = x; b = 0; }
    else if (60 <= h && h < 120) { r = x; g = c; b = 0; } 
    else if (120 <= h && h < 180) { r = 0; g = c; b = x; } 
    else if (180 <= h && h < 240) { r = 0; g = x; b = c; } 
    else if (240 <= h && h < 300) { r = x; g = 0; b = c; } 
    else if (300 <= h && h < 360) { r = c; g = 0; b = x; }
    
    return [r += m,g += m,b += m];
}

function namedColours() {
    return {
        "aliceblue": "0.9411764705882353,0.9725490196078431,1",
        "antiquewhite": "0.9803921568627451,0.9215686274509803,0.8431372549019608",
        "aqua": "0,1,1",
        "aquamarine": "0.4980392156862745,1,0.8313725490196079",
        "azure": "0.9411764705882353,1,1",
        "beige": "0.9607843137254902,0.9607843137254902,0.8627450980392157",
        "bisque": "1,0.8941176470588236,0.7686274509803922",
        "black": "0,0,0",
        "blanchedalmond": "1,0.9215686274509803,0.803921568627451",
        "blue": "0,0,1",
        "blueviolet": "0.5411764705882353,0.16862745098039217,0.8862745098039215",
        "brown": "0.6470588235294118,0.16470588235294117,0.16470588235294117",
        "burlywood": "0.8705882352941177,0.7215686274509804,0.5294117647058824",
        "cadetblue": "0.37254901960784315,0.6196078431372549,0.6274509803921569",
        "chartreuse": "0.4980392156862745,1,0",
        "chocolate": "0.8235294117647058,0.4117647058823529,0.11764705882352941",
        "coral": "1,0.4980392156862745,0.3137254901960784",
        "cornflowerblue": "0.39215686274509803,0.5843137254901961,0.9294117647058824",
        "cornsilk": "1,0.9725490196078431,0.8627450980392157",
        "crimson": "0.8627450980392157,0.0784313725490196,0.23529411764705882",
        "cyan": "0,1,1",
        "darkblue": "0,0,0.5450980392156862",
        "darkcyan": "0,0.5450980392156862,0.5450980392156862",
        "darkgoldenrod": "0.7215686274509804,0.5254901960784314,0.043137254901960784",
        "darkgray": "0.6627450980392157,0.6627450980392157,0.6627450980392157",
        "darkgrey": "0.6627450980392157,0.6627450980392157,0.6627450980392157",
        "darkgreen": "0,0.39215686274509803,0",
        "darkkhaki": "0.7411764705882353,0.7176470588235294,0.4196078431372549",
        "darkmagenta": "0.5450980392156862,0,0.5450980392156862",
        "darkolivegreen": "0.3333333333333333,0.4196078431372549,0.1843137254901961",
        "darkorange": "1,0.5490196078431373,0",
        "darkorchid": "0.6,0.19607843137254902,0.8",
        "darkred": "0.5450980392156862,0,0",
        "darksalmon": "0.9137254901960784,0.5882352941176471,0.47843137254901963",
        "darkseagreen": "0.5607843137254902,0.7372549019607844,0.5607843137254902",
        "darkslateblue": "0.2823529411764706,0.23921568627450981,0.5450980392156862",
        "darkslategray": "0.1843137254901961,0.30980392156862746,0.30980392156862746",
        "darkslategrey": "0.1843137254901961,0.30980392156862746,0.30980392156862746",
        "darkturquoise": "0,0.807843137254902,0.8196078431372549",
        "darkviolet": "0.5803921568627451,0,0.8274509803921568",
        "deeppink": "1,0.0784313725490196,0.5764705882352941",
        "deepskyblue": "0,0.7490196078431373,1",
        "dimgray": "0.4117647058823529,0.4117647058823529,0.4117647058823529",
        "dimgrey": "0.4117647058823529,0.4117647058823529,0.4117647058823529",
        "dodgerblue": "0.11764705882352941,0.5647058823529412,1",
        "firebrick": "0.6980392156862745,0.13333333333333333,0.13333333333333333",
        "floralwhite": "1,0.9803921568627451,0.9411764705882353",
        "forestgreen": "0.13333333333333333,0.5450980392156862,0.13333333333333333",
        "fuchsia": "1,0,1",
        "gainsboro": "0.8627450980392157,0.8627450980392157,0.8627450980392157",
        "ghostwhite": "0.9725490196078431,0.9725490196078431,1",
        "gold": "1,0.8431372549019608,0",
        "goldenrod": "0.8549019607843137,0.6470588235294118,0.12549019607843137",
        "gray": "0.5019607843137255,0.5019607843137255,0.5019607843137255",
        "grey": "0.5019607843137255,0.5019607843137255,0.5019607843137255",
        "green": "0,0.5019607843137255,0",
        "greenyellow": "0.6784313725490196,1,0.1843137254901961",
        "honeydew": "0.9411764705882353,1,0.9411764705882353",
        "hotpink": "1,0.4117647058823529,0.7058823529411765",
        "indianred": "0.803921568627451,0.3607843137254902,0.3607843137254902",
        "indigo": "0.29411764705882354,0,0.5098039215686274",
        "ivory": "1,1,0.9411764705882353",
        "khaki": "0.9411764705882353,0.9019607843137255,0.5490196078431373",
        "lavender": "0.9019607843137255,0.9019607843137255,0.9803921568627451",
        "lavenderblush": "1,0.9411764705882353,0.9607843137254902",
        "lawngreen": "0.48627450980392156,0.9882352941176471,0",
        "lemonchiffon": "1,0.9803921568627451,0.803921568627451",
        "lightblue": "0.6784313725490196,0.8470588235294118,0.9019607843137255",
        "lightcoral": "0.9411764705882353,0.5019607843137255,0.5019607843137255",
        "lightcyan": "0.8784313725490196,1,1",
        "lightgoldenrodyellow": "0.9803921568627451,0.9803921568627451,0.8235294117647058",
        "lightgray": "0.8274509803921568,0.8274509803921568,0.8274509803921568",
        "lightgrey": "0.8274509803921568,0.8274509803921568,0.8274509803921568",
        "lightgreen": "0.5647058823529412,0.9333333333333333,0.5647058823529412",
        "lightpink": "1,0.7137254901960784,0.7568627450980392",
        "lightsalmon": "1,0.6274509803921569,0.47843137254901963",
        "lightseagreen": "0.12549019607843137,0.6980392156862745,0.6666666666666666",
        "lightskyblue": "0.5294117647058824,0.807843137254902,0.9803921568627451",
        "lightslategray": "0.4666666666666667,0.5333333333333333,0.6",
        "lightslategrey": "0.4666666666666667,0.5333333333333333,0.6",
        "lightsteelblue": "0.6901960784313725,0.7686274509803922,0.8705882352941177",
        "lightyellow": "1,1,0.8784313725490196",
        "lime": "0,1,0",
        "limegreen": "0.19607843137254902,0.803921568627451,0.19607843137254902",
        "linen": "0.9803921568627451,0.9411764705882353,0.9019607843137255",
        "magenta": "1,0,1",
        "maroon": "0.5019607843137255,0,0",
        "mediumaquamarine": "0.4,0.803921568627451,0.6666666666666666",
        "mediumblue": "0,0,0.803921568627451",
        "mediumorchid": "0.7294117647058823,0.3333333333333333,0.8274509803921568",
        "mediumpurple": "0.5764705882352941,0.4392156862745098,0.8470588235294118",
        "mediumseagreen": "0.23529411764705882,0.7019607843137254,0.44313725490196076",
        "mediumslateblue": "0.4823529411764706,0.40784313725490196,0.9333333333333333",
        "mediumspringgreen": "0,0.9803921568627451,0.6039215686274509",
        "mediumturquoise": "0.2823529411764706,0.8196078431372549,0.8",
        "mediumvioletred": "0.7803921568627451,0.08235294117647059,0.5215686274509804",
        "midnightblue": "0.09803921568627451,0.09803921568627451,0.4392156862745098",
        "mintcream": "0.9607843137254902,1,0.9803921568627451",
        "mistyrose": "1,0.8941176470588236,0.8823529411764706",
        "moccasin": "1,0.8941176470588236,0.7098039215686275",
        "navajowhite": "1,0.8705882352941177,0.6784313725490196",
        "navy": "0,0,0.5019607843137255",
        "oldlace": "0.9921568627450981,0.9607843137254902,0.9019607843137255",
        "olive": "0.5019607843137255,0.5019607843137255,0",
        "olivedrab": "0.4196078431372549,0.5568627450980392,0.13725490196078433",
        "orange": "1,0.6470588235294118,0",
        "orangered": "1,0.27058823529411763,0",
        "orchid": "0.8549019607843137,0.4392156862745098,0.8392156862745098",
        "palegoldenrod": "0.9333333333333333,0.9098039215686274,0.6666666666666666",
        "palegreen": "0.596078431372549,0.984313725490196,0.596078431372549",
        "paleturquoise": "0.6862745098039216,0.9333333333333333,0.9333333333333333",
        "palevioletred": "0.8470588235294118,0.4392156862745098,0.5764705882352941",
        "papayawhip": "1,0.9372549019607843,0.8352941176470589",
        "peachpuff": "1,0.8549019607843137,0.7254901960784313",
        "peru": "0.803921568627451,0.5215686274509804,0.24705882352941178",
        "pink": "1,0.7529411764705882,0.796078431372549",
        "plum": "0.8666666666666667,0.6274509803921569,0.8666666666666667",
        "powderblue": "0.6901960784313725,0.8784313725490196,0.9019607843137255",
        "purple": "0.5019607843137255,0,0.5019607843137255",
        "red": "1,0,0",
        "rebeccapurple": "0.4,0.2,0.6",
        "rosybrown": "0.7372549019607844,0.5607843137254902,0.5607843137254902",
        "royalblue": "0.2549019607843137,0.4117647058823529,0.8823529411764706",
        "saddlebrown": "0.5450980392156862,0.27058823529411763,0.07450980392156863",
        "salmon": "0.9803921568627451,0.5019607843137255,0.4470588235294118",
        "sandybrown": "0.9568627450980393,0.6431372549019608,0.3764705882352941",
        "seagreen": "0.1803921568627451,0.5450980392156862,0.3411764705882353",
        "seashell": "1,0.9607843137254902,0.9333333333333333",
        "sienna": "0.6274509803921569,0.3215686274509804,0.17647058823529413",
        "silver": "0.7529411764705882,0.7529411764705882,0.7529411764705882",
        "skyblue": "0.5294117647058824,0.807843137254902,0.9215686274509803",
        "slateblue": "0.41568627450980394,0.35294117647058826,0.803921568627451",
        "slategray": "0.4392156862745098,0.5019607843137255,0.5647058823529412",
        "slategrey": "0.4392156862745098,0.5019607843137255,0.5647058823529412",
        "snow": "1,0.9803921568627451,0.9803921568627451",
        "springgreen": "0,1,0.4980392156862745",
        "steelblue": "0.27450980392156865,0.5098039215686274,0.7058823529411765",
        "tan": "0.8235294117647058,0.7058823529411765,0.5490196078431373",
        "teal": "0,0.5019607843137255,0.5019607843137255",
        "thistle": "0.8470588235294118,0.7490196078431373,0.8470588235294118",
        "tomato": "1,0.38823529411764707,0.2784313725490196",
        "turquoise": "0.25098039215686274,0.8784313725490196,0.8156862745098039",
        "violet": "0.9333333333333333,0.5098039215686274,0.9333333333333333",
        "wheat": "0.9607843137254902,0.8705882352941177,0.7019607843137254",
        "white": "1,1,1",
        "whitesmoke": "0.9607843137254902,0.9607843137254902,0.9607843137254902",
        "yellow": "1,1,0",
        "yellowgreen": "0.6039215686274509,0.803921568627451,0.19607843137254902"
    };
}


